pkg/networks/usernet/gvproxy.go

```diff
@@ -0,0 +1,229 @@
+	if err != nil {
+		return err
+	}
+
+	ln, err := transport.Listen(fmt.Sprintf("unix://%s", opts.Endpoint))
+	if err != nil {
+		return err
+	}
+	httpServe(ctx, g, ln, vn.Mux())
+
+	if opts.QemuSocket != "" {
+		err = listenQEMU(ctx, vn)
+		if err != nil {
+			return err
+		}
+	}
+	if opts.FdSocket != "" {
+		err = listenFD(ctx, vn)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func listenQEMU(ctx context.Context, vn *virtualnetwork.VirtualNetwork) error {
+	listener, err := net.Listen("unix", opts.QemuSocket)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		defer listener.Close()
+		for {
+			conn, err := listener.Accept()
+			if err != nil {
+				logrus.Error("QEMU accept failed", err)
+			}
+
+			go func() {
+				err = vn.AcceptQemu(ctx, conn)
+				if err != nil {
+					logrus.Error("QEMU connection closed with error", err)
+				}
+				conn.Close()
+			}()
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				continue
+			}
+		}
+
+	}()
+
+	return nil
+}
+
+func listenFD(ctx context.Context, vn *virtualnetwork.VirtualNetwork) error {
+	listener, err := net.Listen("unix", opts.FdSocket)
+	if err != nil {
+		return err
+	}
+
+	go func() {
+		defer listener.Close()
+		for {
+			conn, err := listener.Accept()
+			if err != nil {
+				logrus.Error("FD accept failed", err)
+			}
+
+			files, err := fd.Get(conn.(*net.UnixConn), 1, []string{"client"})
+			if err != nil {
+				logrus.Error("Failed to get FD via socket", err)
+			}
+
+			if len(files) != 1 {
+				logrus.Error("Invalid number of fd in response", err)
+			}
+			fileConn, err := net.FileConn(files[0])
+			if err != nil {
+				logrus.Error("Error in FD Socket", err)
+			}
+			files[0].Close()
+
+			go func() {
+				err = vn.AcceptBess(ctx, fileConn)
+				if err != nil {
+					logrus.Error("FD connection closed with error", err)
+				}
+				defer fileConn.Close()
+			}()
+			select {
+			case <-ctx.Done():
+				return
+			default:
+				continue
+			}
+		}
+	}()
+
+	return nil
+}
+
+func httpServe(ctx context.Context, g *errgroup.Group, ln net.Listener, mux http.Handler) {
+	g.Go(func() error {
+		<-ctx.Done()
+		return ln.Close()
+	})
+	g.Go(func() error {
+		s := &http.Server{
+			Handler:      mux,
+			ReadTimeout:  10 * time.Second,
+			WriteTimeout: 10 * time.Second,
+		}
+		err := s.Serve(ln)
+		if err != nil {
+			if err != http.ErrServerClosed {
+				return err
+			}
+			return err
+		}
+		return nil
+	})
+}
+
+func searchDomains() []string {
+	if runtime.GOOS != "windows" {
+		f, err := os.Open("/etc/resolv.conf")
+		if err != nil {
+			logrus.Errorf("open file error: %v", err)
+			return nil
+		}
+		defer f.Close()
+		sc := bufio.NewScanner(f)
+		searchPrefix := "search "
+		for sc.Scan() {
+			if strings.HasPrefix(sc.Text(), searchPrefix) {
+				searchDomains := strings.Split(strings.TrimPrefix(sc.Text(), searchPrefix), " ")
+				logrus.Debugf("Using search domains: %v", searchDomains)
+				return searchDomains
+			}
```

Comment:
Can we have a unit test for this function?

Reply:
Done